libportability-gfx/src/impls.rs

```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
 }
 #[inline]
 pub extern "C" fn gfxCreateGraphicsPipelines(
-    device: VkDevice,
+    gpu: VkDevice,
     pipelineCache: VkPipelineCache,
     createInfoCount: u32,
     pCreateInfos: *const VkGraphicsPipelineCreateInfo,
-    pAllocator: *const VkAllocationCallbacks,
+    _pAllocator: *const VkAllocationCallbacks,
     pPipelines: *mut VkPipeline,
 ) -> VkResult {
-    unimplemented!()
+    assert!(pipelineCache.is_null());
+
+    let infos = unsafe {
+        slice::from_raw_parts(pCreateInfos, createInfoCount as _)
+    };
+
+    const NUM_SHADER_STAGES: usize = 5;
+    let mut shader_stages = Vec::with_capacity(infos.len() * NUM_SHADER_STAGES);
+
+    // Collect all information which we will borrow later. Need to work around
+    // the borrow checker here.
+    // TODO: try to refactor it once we have a more generic API
+    for info in infos {
+        let stages = unsafe {
+            slice::from_raw_parts(info.pStages, info.stageCount as _)
+        };
+
+        for stage in stages {
+            let name = unsafe { CStr::from_ptr(stage.pName).to_owned() };
+            let specialization = unsafe { stage
+                .pSpecializationInfo
+                .as_ref()
+                .map(|specialization| {
+                    let data = slice::from_raw_parts(
+                        specialization.pData,
+                        specialization.dataSize as _,
+                    );
+                    let entries = slice::from_raw_parts(
+                        specialization.pMapEntries,
+                        specialization.mapEntryCount as _,
+                    );
+
+                    entries.
+                        into_iter()
+                        .map(|entry| {
+                            // Currently blocked due to lack of specialization type knowledge
+                            unimplemented!()
+                        })
+                        .collect::<Vec<pso::Specialization>>()
+                })
+                .unwrap_or(vec![])
+            };
+
+            shader_stages.push((
+                name.into_string().unwrap(),
+                specialization,
+            ));
+        }
+    }
+
+    let mut cur_shader_stage = 0;
+
+    let descs = infos.into_iter().map(|info| {
+        // TODO: handle dynamic states and viewports
+
+        let shaders = {
+            let mut set: pso::GraphicsShaderSet<_> = unsafe { mem::zeroed() };
+
+            let stages = unsafe {
+                slice::from_raw_parts(info.pStages, info.stageCount as _)
+            };
+
+            for stage in stages {
+                use super::VkShaderStageFlagBits::*;
+
+                let (ref name, ref specialization) = shader_stages[cur_shader_stage];
+                cur_shader_stage += 1;
+
+                let entry_point = pso::EntryPoint {
+                    entry: &name,
+                    module: stage.module.deref(),
```

Comment:
note: you can achieve the same without importing `Deref` by doing `&*stage.module`

Reply:
I think it was `&**stage` for these deref cases, not sure what's nicer..

Reply:
Arguably, this is still more explicit than `x.deref()` :)